File: 33_search-in-rotated-sorted-array.py
class Solution:
  def search(self, nums: List[int], target: int) -> int:

    def find(l:int, r:int) -> int:
      if l > r:
        return -1
      nonlocal nums, target
      mid = (l + r) >> 1
      if nums[mid] == target:
        return mid
  
      if nums[mid] < nums[r]: # sorted
        if nums[mid] <= target and target <= nums[r]:
          return find(mid+1, r)
        else:
          return find(l, mid-1)
      else:
        if nums[l] <= target and nums[mid] >= target: 
          return find(l, mid-1)
        else:
          return find(mid+1, r)

    return find(0, len(nums)-1)

File: BinarySearch/33-search-in-rotated-sorted-array.cpp
/******************************************
 * author      : utoppia
 * description : solution for leetcode
 * updated at  : 2024-01-12 23:54:38
 * filename    : 33-search-in-rotated-sorted-array.cpp
 * language    : cpp
 * editor      : Emacs 27.1
 *****************************************/
#include <algorithm>
#include <bits/stdc++.h>
#include <ios>
#include <iostream>
#include <set>
#include <utility>
#include <vector>

#define vall(v) v.begin(), v.end()
#define pir pair<int, int>

#define pb push_back
#define eb emplace_back
#define fi first
#define snd second
#define sz size

#define ll long long
#define nMax 200005

#define MOD 998244353
using namespace std;

class Solution {

public:
  int search(vector<int> &nums, int target) {
    int n = nums.size();
    int l = 0, r = n - 1;
    while (l <= r) {
      int m = (l + r) >> 1;
      if (nums[m] == target)
        return m;
      if (m + 1 < n && nums[m + 1] <= nums[r]) {
        if (nums[m + 1] <= target && target <= nums[r]) {
          l = m + 1;
        } else {
          r = m - 1;
        }
      } else {
        if (nums[l] <= target && target <= nums[m]) {
          r = m - 1;
        } else {
          l = m + 1;
        }
      }
    }
    return -1;
  }
};
